### ISSUE TITLE
Feature Request: Set multiplication

### ISSUE THREAD

**Issue by akvlad** [open-policy-agent/opa#3800]
## Expected Behavior
It would be great if there is a set multiplication function in Rego, something like:

```
a := {"a", "b"}
b := {"c", "d"}
c := multiply(a,b)
#c == {["a", "c"], ["a", "d"], ["b", "c"], ["b", "d"]}
```
## Actual Behavior
Now it is achievable via a rego function:
```
multiply = res {
    res := {[_a, _b] |
        a[_a]
        b[_b]
    }
}
```
But the native acceleration (like for union, intersect and subtract) is very welcome.

**Comment by srenatus** [comment]
Sounds like a good addition. What's your use case, if you can share that?

**Comment by akvlad** [comment]
My case is very similar to the described one but the "a" set may have up to 5000 strings.
The sets are uploaded from the database to "data" via OPAL / restful API. Then I have to multiply them. Then do some unions/subtractions. 
But the case is very same. And the multiply function is the same.

**Comment by srenatus** [comment]
Hmm interesting. And you need the whole product set...? I'm asking because there might be ways around that. I.e. if you wanted to check if an element `x` is the product set, you could also do something like
```rego
in_product(a, xs, ys) {
  xs[a]
  ys[a]
}
```

**Comment by akvlad** [comment]
Yes. Because after the multiplication I do "group by" in the python wrapper. (I'm going to make another feature request for the "group by"):
```
res = dict()
for e in product_set:
  if e[0] not in res:
    res[e[0]] = []
  res[e[0]].append(e[1])
```
And "res" is the answer for some RESTful requests. Not the most often requests, actually, but the performance improvement in any step would be very useful.

**Comment by andriyfomenko** [comment]
I'm working with Vlad on his project, and yet, he indeed needs to get a full set out of OPA in this use case.

We have the pointed "access to only this particular combination" in other cases and these work very fast, but the product is slow to get.

**Comment by akvlad** [comment]
Let me show the whole task we are trying to solve:
```
s_11 := {"o1", "o2"}
s_12 := {"a1", "a2"}

s_21 := {"o2", "o3"}
s_22 := {"a2", "a3"}

s_31 := {"o3", "o4"}
s_32 := {"a2", "a4"}

res := union({multiply(s_11, s_12), multiply(s_21, s_22)}) - multiply(s_31, s_32)
```
I tried several approaches to resolve this task and this seems the fastest one. Now it takes about 0.3s for a resulting set for 50000 entries. It is OK for us now, but improvements are highly appreciated.

**Comment by srenatus** [comment]
Thanks for sharing. I think this is a good built-in to add for anyone wanting to get into contributing to OPA. Would that include you? 😃

**Comment by akvlad** [comment]
I promise nothing, but I have forked. ;)

**Comment by anderseknert** [comment]
The [PR](https://github.com/open-policy-agent/opa/pull/3815) to implement this was abandonded, so closing this issue for now. If anyone else feels like this would help solve a problem they're facing, and/or would like to work on adding it, let me know and we'll have it re-opened.
